### DISCUSSION TITLE
Fetching contacts fails

### DISCUSSION THREAD

**Discussion by svdvonde** [yu-i-i/overleaf-cep#4] [Q&A] [answered]
Hi,

In Overleaf's share menu, I'd like to make all LDAP users available as contact (findable by name or email) for all other LDAP users. Apologies if I'm making basic configuration mistakes, but this is my attempt at configuring this contact relation: 

```text
OVERLEAF_LDAP_SEARCH_BASE=ou=People,dc=vub,dc=ac,dc=be
OVERLEAF_LDAP_SEARCH_FILTER=(|(uid={{username}})(mail={{username}}))
OVERLEAF_LDAP_EMAIL_ATT=mail
OVERLEAF_LDAP_FIRST_NAME_ATT=givenName
OVERLEAF_LDAP_LAST_NAME_ATT=sn
OVERLEAF_LDAP_NAME_ATT=cn
OVERLEAF_LDAP_SEARCH_ATTRIBUTES=["uid", "sn", "givenName", "mail"]

OVERLEAF_LDAP_CONTACTS_FILTER=(objectClass=person)
```

When I try this out, an error is thrown by [this line of code](https://github.com/yu-i-i/overleaf-cep/blob/ldap-tc/services/web/modules/ldap-authentication/app/src/LdapContacts.js#L53) in the application:
```
web           | {"name":"web","hostname":"b261c454490f","pid":203,"level":50,"err":{"message":"ldapUser[attEmail]?.toLowerCase is not a function","name":"TypeError","stack":"TypeError: ldapUser[attEmail]?.toLowerCase is not a function\n    at /overleaf/services/web/modules/ldap-authentication/app/src/LdapContacts.js:53:36\n    at Array.reduce (<anonymous>)\n    at fetchLdapContacts (/overleaf/services/web/modules/ldap-authentication/app/src/LdapContacts.js:44:37)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async getContacts (/overleaf/services/web/modules/ldap-authentication/index.js:20:33)","info":{}},"responseTimeMs":49,"req":{"method":"GET","url":"/user/contacts","remoteAddress":"127.0.0.1","headers":{"referer":"http://myldap-server/project/66ed882aedcaf8b4a66179ec","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0"}},"res":{"statusCode":500,"headers":{"content-length":"796"}},"msg":"GET /user/contacts","time":"2024-09-20T14:51:11.592Z","v":0}
```

**Comment by yu-i-i** [comment, marked as answer]
Try the following filter instead:
```
OVERLEAF_LDAP_CONTACTS_FILTER=(&(objectClass=person)(mail=*))
```
It seems that one of the entries in your search results has an undefined mail attribute, and the code incorrectly assumed that this wouldn’t happen. I'll resolve this by skipping such entries from the search results. However, please note that having a mail attribute is essential — all users and contacts must have an email address defined, as it is used as an ID in the Overleaf database.

**Reply by svdvonde** [reply]
Heh. I'll talk to our LDAP admin about that. I believe that there are some accounts which do not belong to human beings.

I'll try out the filter tomorrow, and will try to find these users without a mail attribute. Thank you again!

**Reply by svdvonde** [reply]
I tried out the modified LDAP search filter, but the same error is thrown every time when the share menu is opened.

**Reply by yu-i-i** [reply]
Thanks for your comments. I found a bug when an LDAP user has multiple email addresses. This is probably the issue you're facing. It was fixed in the last commit.

**Reply by svdvonde** [reply]
Thank you! That has resolved this particular issue, indeed! Lookup is behaving a little strangely though. When I enter 1 character, then backspace, the list of contacts does not seem to be emptying properly, and every time I enter 1 character and press backspace the list of contacts seems to keep growing.

Below I'll attach a screenshot that is the result of me "entering  1 character and pressing backspace" a couple of times, followed by me clicking outside the email box. For some reason this turns the perpetually growing list of contacts into a bullet list. I've blacked out the email addresses of these users (which happened to persist in the list), but it's continuously the same (few?) accounts.

![Untitled](https://github.com/user-attachments/assets/a2d0292e-a078-4c0c-a16a-e9db5c5f0a91)

**Reply by yu-i-i** [reply]
>  but it's continuously the same (few?) accounts.

Sorry. I didn't pay attention to this, when I wrote you the previous reply. This is rather strange. I don't have such behavior. Could you please try it with another browser?

**Reply by svdvonde** [reply]
Yep, seems to be the same 3 users every time. Some observations:

- These 3 contacts seem to be suggested in autocomplete for every search which has at least 1 result. If the query gets too specific (matches no contact), then the autocomplete does not suggest any contacts, but clicking outside the box makes the shown bullet list appear.
- The hover behaviour of these autocomplete suggestions seems broken. "normal results" (that I would expect) get a green background when hovered, and these 3 contacts do not get a green background on hover. Furthermore I cannot click these 3 suggestions to select the contact - when I do click them, bullet lists appears.
- Every time I clear the sharing box and search for a different contact, the 3 contact are appended to the list again together with the last time I searched for a contact. So if I "search for a user + clear the box" x2, then 6 broken contacts are shown. Seems to be at the front of the autocomplete too (top search suggestions) despite them not matching the query. 
- Tested on the latest Edge browser (Chromium), as well as Firefox. Same behaviour.

**Reply by yu-i-i** [reply]
Can you safely delete the user whose contacts exhibit this behavior from mongo database? If so, could you recreate the user and observe what happens? If the behavior remains the same, I would suggest examining the contacts behavior in the mainstream version, without ldap. Unfortunately, I cannot reproduce the bug you have.

**Reply by svdvonde** [reply]
Our self-hosted version of Overleaf is not (yet) being used by users, so the only user is me, and I can destroy the database if I want.

There seems to be something particular about the LDAP config of these 3 users. I can exclude them in the contacts filter:

```
OVERLEAF_LDAP_CONTACTS_FILTER=(&(objectClass=person)(mail=*)(!(uid=problem-user-1))(!(uid=problem-user-2))(!(uid=problem-user-3)))
```

This basically resolves the issue, and the sharing menu works as expected. Tomorrow I'll try to find out what makes these 3 users different from the others. Maybe it can help track down a very specific bug.

**Reply by svdvonde** [reply]
I looked at the attributes of these 3 users compared to others, and I could not find a common denominator that somehow made them different. The problem shall remain a mystery 🙂 In my case, it is not longer an issue. When I additionally filtered for active accounts (which is desirable anyway), the problematic users were also excluded.

**Reply by yu-i-i** [reply]
I’m not a fan of such mysteries, so here's one last attempt :) Is it possible that the names or emails of these users contain unprintable characters? It shouldn't be an issue, but you never know. Would you be able to provide the ldapsearch output for one of these users, with any sensitive data redacted?

**Reply by svdvonde** [reply]
The following is the output of `ldapsearch` of 1 of the problematic accounts, the only 'redacted' info is the local part of `mail` which was a normal email address with normal `[a-z]` characters. 

```text
# moves, People, vub.ac.be
dn: uid=moves,ou=People,dc=vub,dc=ac,dc=be
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: posixAccount
objectClass: shadowAccount
objectClass: sambaSamAccount
cn: User MOVES
sn: MOVES
givenName: User
uid: moves
uidNumber: 1108
gidNumber: 1110
homeDirectory: /home/moves
loginShell: /bin/bash
gecos: User MOVES
mail: xxxxxxxx@vub.ac.be
sambaLogonTime: 0
sambaLogoffTime: 2147483647
sambaPwdCanChange: 0
displayName: User MOVES
sambaSID: S-1-5-21-3904260993-2295595841-432895439-3216
shadowExpire: 10000015172
shadowMax: 9999999999
sambaKickoffTime: 864001310906676
sambaAcctFlags: [DU]
```

Compare this to the `ldapsearch` output of my account which is non-problematic (with redacted email .. for bots or something): 

```text
# svdvonde, People, vub.ac.be
dn: uid=svdvonde,ou=People,dc=vub,dc=ac,dc=be
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: posixAccount
objectClass: shadowAccount
objectClass: inetOrgPerson
objectClass: sambaSamAccount
cn: Sam Van den Vonder
sn: Van den Vonder
uid: svdvonde
uidNumber: 1200
gidNumber: 1193
homeDirectory: /home/svdvonde
loginShell: /bin/bash
gecos: Sam Van den Vonder
givenName: Sam
sambaLogonTime: 0
sambaLogoffTime: 2147483647
sambaPwdCanChange: 0
displayName: Sam Van den Vonder
sambaSID: S-1-5-21-3904260993-2295595841-432895439-3400
sambaPrimaryGroupSID: S-1-5-21-3904260993-2295595841-432895439-3387
sambaLogonScript: logon.bat
sambaHomePath: \\SOFT\svdvonde
sambaHomeDrive: H:
sambaAcctFlags: [U]
shadowExpire: 10000017041
shadowMax: 9999999999
sambaKickoffTime: 864001472392192
roomNumber: F.10.730
mail: xxxxxxxx@vub.ac.be
```

I diff'd the two records (as well as for the other 2 problematic users), and I cannot identify anything unusual. If you have any theories I'd be happy to check it against the 2 other problematic users as well.

**Comment by yu-i-i** [comment]
I cannot fix this; my code only builds a list of LDAP contacts, while displaying the contacts is managed by the main code. Just a small note: when you type a letter in the empty input field, it searches for contacts containing that letter. After pressing backspace (input field is empty again), it shows the full list of all contacts, though I have a window with a scrollbar where I can only see 8 contacts at a time.

I also noticed something in the upstream code mentioning a limit of 50 contacts. Probably, the code may show you 50 contacts initially, and then another 50 after the next backspace. I'm not sure—just a guess.

Additionally, I have a general thought: typically, a person has a limited number of potential collaborators, often from the same department. I believe it would be beneficial to structure the LDAP database so that people from the same department share the same group ID or something similar. This way, a filter could be set up to show each user only those with the same property, keeping the contacts list manageable.
